prov/efa/src/rxr/rxr_ep.c

```diff
@@ -1120,44 +1120,28 @@ static ssize_t rxr_ep_cancel(fid_t fid_ep, void *context)
  */
 static int efa_set_fi_hmem_p2p_opt(struct efa_ep *efa_ep, int opt)
 {
-	struct efa_hmem_info *hmem_info;
+	int i, ret;
 
-	hmem_info = efa_ep->domain->hmem_info;
-
-	switch (opt) {
+	/* FI_HMEM interfaces to check against */
+	static enum fi_hmem_iface ifaces[] = {
+		FI_HMEM_CUDA,
+		FI_HMEM_NEURON,
+		FI_HMEM_SYNAPSEAI
+	};
+	
 	/*
-	 * TODO: support the other options. We can only support ENABLED
-	 * and PREFERRED when p2p is available. DISABLED is not
-	 * supported yet.
+	 * Check the opt's validity against the first initialized FI_HMEM interface
+	 * encountered from the list above
 	 */
-	case FI_HMEM_P2P_REQUIRED:
-	case FI_HMEM_P2P_ENABLED:
-	case FI_HMEM_P2P_PREFERRED:
-		if (hmem_info[FI_HMEM_CUDA].initialized &&
-		    hmem_info[FI_HMEM_CUDA].p2p_supported) {
-			efa_ep->hmem_p2p_opt = opt;
-		} else if (hmem_info[FI_HMEM_NEURON].initialized &&
-			   hmem_info[FI_HMEM_NEURON].p2p_supported) {
-			/*
-			 * Neuron requires p2p support and supports no
-			 * other modes.
-			 */
-			if (opt != FI_HMEM_P2P_REQUIRED)
-				return -FI_EOPNOTSUPP;
-			efa_ep->hmem_p2p_opt = FI_HMEM_P2P_REQUIRED;
-		} else {
-			return -FI_EOPNOTSUPP;
+	for (i = 0; i < ((sizeof ifaces) / (sizeof (enum fi_hmem_iface))); ++i) {
+		ret = efa_hmem_validate_p2p_opt(efa_ep->domain, ifaces[i], opt);
+		if (ret != FI_ENODATA) {
+			if (!ret)
+				efa_ep->hmem_p2p_opt = opt;
+			return -ret;
 		}
-
-		break;
-	case FI_HMEM_P2P_DISABLED:
-		return -FI_EOPNOTSUPP;
-		break;
-	default:
-		return -FI_EINVAL;
 	}
-
-	return 0;
+	return -FI_EINVAL;
```

Comment:
If no hmem interface is enabled, we return -FI_EINVAL, but perhaps that would be considered a success as well?

Reply:
Shouldn't a `FI_OPT_FI_HMEM_P2P` option technically be invalid if there are no available `FI_HMEM` interfaces?